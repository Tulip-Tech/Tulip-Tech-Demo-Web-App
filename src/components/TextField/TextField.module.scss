@use 'src/styles/variables';
@use 'src/styles/theme';

.textField {
  width: 100%;
  margin-bottom: 8px;

  &.error {
    .container {
      border-color: theme.$text-field-error-color;
    }
  }

  &.disabled {
    .container {
      opacity: 0.7;
    }
  }

  &.leftControl .input {
    padding-left: 0;
  }

  &.rightControl .input {
    padding-right: 0;
  }

  &:hover {
    &:not(.disabled) {
      .container {
        background-color: theme.$text-field-hover-bg-color;
        border-color: theme.$text-field-hover-border-color;
      }
    }
  }
}

.label {
  display: block;
  margin-bottom: 4px;
  font-weight: var(--body-font-weight-bold);
  text-align: left;

  > span {
    float: right;
    color: rgba(variables.$white, 0.7);
    font-weight: normal;
  }
}

.control > div {
  width: 48px;
  height: 48px;
}

.container {
  display: flex;
  width: 100%;

  color: theme.$text-field-resting-color;

  background-color: theme.$text-field-bg-color;
  border: 1px solid theme.$text-field-resting-border-color;
  border-radius: 4px;
  transition: border 0.2s ease;

  &:focus-within {
    color: theme.$text-field-active-color;
    border-color: theme.$text-field-active-border-color;
  }
}

.input {
  width: 100%;
  min-height: 48px;
  padding: 14px 16px;
  color: inherit;
  font-size: 16px;
  line-height: 18px;
  background: transparent;
  border: none;
  outline: none;
  appearance: none;

  &[type='password'] {
    &::-ms-reveal {
      display: none;
    }
  }
}
