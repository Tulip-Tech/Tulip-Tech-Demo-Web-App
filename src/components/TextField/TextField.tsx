import React, { RefObject } from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import styles from './TextField.module.scss';

import HelperText from '#components/HelperText/HelperText';
import { testId as getTestId } from '#src/utils/common';
import useOpaqueId from '#src/hooks/useOpaqueId';

type InputProps = Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, 'id' | 'ref' | 'className'>;
type TextAreaProps = Omit<React.DetailedHTMLProps<React.TextareaHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement>, 'id' | 'ref' | 'className'>;

type InputOrTextAreaProps =
  | ({ multiline?: never; inputRef?: RefObject<HTMLInputElement>; textAreaRef?: never } & InputProps)
  | ({ multiline: true; inputRef?: never; textAreaRef?: RefObject<HTMLTextAreaElement> } & TextAreaProps);

type Props = {
  className?: string;
  label?: string;
  helperText?: React.ReactNode;
  leftControl?: React.ReactNode;
  rightControl?: React.ReactNode;
  error?: boolean;
  editing?: boolean;
  testId?: string;
  multiline?: boolean;
} & InputOrTextAreaProps;

const TextField: React.FC<Props> = ({
  className,
  label,
  error,
  helperText,
  leftControl,
  rightControl,
  editing = true,
  testId,
  inputRef,
  textAreaRef,
  multiline,
  ...inputProps
}: Props) => {
  const id = useOpaqueId('text-field', inputProps.name);
  const { t } = useTranslation('common');

  const isInputOrTextArea = (item: unknown): item is InputOrTextAreaProps => !!item && typeof item === 'object';
  const isTextArea = (item: unknown): item is TextAreaProps => isInputOrTextArea(item) && !!multiline;

  const textFieldClassName = classNames(
    styles.textField,
    {
      [styles.error]: error,
      [styles.disabled]: inputProps.disabled,
      [styles.leftControl]: !!leftControl,
      [styles.rightControl]: !!rightControl,
    },
    className,
  );

  return (
    <div className={textFieldClassName} data-testid={getTestId(testId)}>
      <label htmlFor={id} className={styles.label}>
        {label}
        {!inputProps.required && editing ? <span>{t('optional')}</span> : null}
      </label>
      {editing ? (
        <div className={styles.container}>
          {leftControl ? <div className={styles.control}>{leftControl}</div> : null}
          {isTextArea(inputProps) ? (
            <textarea id={id} className={styles.input} rows={3} ref={textAreaRef} {...inputProps} />
          ) : (
            <input id={id} className={styles.input} type={'text'} ref={inputRef} {...inputProps} />
          )}
          {rightControl ? <div className={styles.control}>{rightControl}</div> : null}
        </div>
      ) : (
        <p>{inputProps.value}</p>
      )}
      <HelperText error={error}>{helperText}</HelperText>
    </div>
  );
};

export default TextField;
