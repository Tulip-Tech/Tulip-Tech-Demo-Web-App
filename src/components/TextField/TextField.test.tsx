import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import TextField from './TextField';

describe('<TextField>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<TextField label="Label" placeholder="Placeholder" name="name" value="" onChange={vi.fn()} />);

    expect(container).toMatchSnapshot();
  });

  test('renders and matches multiline snapshot', () => {
    const { container } = render(<TextField label="Label" placeholder="Placeholder" name="name" value="" onChange={vi.fn()} multiline={true} />);

    expect(container).toMatchSnapshot();
  });

  test('triggers an onChange event when the input value changes', () => {
    const onChange = vi.fn();
    const { getByPlaceholderText } = render(<TextField value="" onChange={onChange} placeholder="Enter your name" />);

    fireEvent.change(getByPlaceholderText('Enter your name'), { target: { value: 'John Doe' } });

    expect(onChange).toBeCalled();
  });

  test('shows the helper text below the input', () => {
    const { queryByText } = render(<TextField value="" onChange={vi.fn()} helperText="Assertive text" />);

    expect(queryByText('Assertive text')).toBeDefined();
  });
});
