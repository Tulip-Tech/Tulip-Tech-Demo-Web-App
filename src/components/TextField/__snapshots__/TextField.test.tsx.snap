// Vitest Snapshot v1

exports[`<TextField> > renders and matches multiline snapshot 1`] = `
<div>
  <div
    class="_textField_e16c1b"
  >
    <label
      class="_label_e16c1b"
      for="text-field_1235_name"
    >
      Label
      <span>
        optional
      </span>
    </label>
    <div
      class="_container_e16c1b"
    >
      <textarea
        class="_input_e16c1b"
        id="text-field_1235_name"
        name="name"
        placeholder="Placeholder"
        rows="3"
      />
    </div>
  </div>
</div>
`;

exports[`<TextField> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_textField_e16c1b"
  >
    <label
      class="_label_e16c1b"
      for="text-field_1235_name"
    >
      Label
      <span>
        optional
      </span>
    </label>
    <div
      class="_container_e16c1b"
    >
      <input
        class="_input_e16c1b"
        id="text-field_1235_name"
        name="name"
        placeholder="Placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
